#38044 API design: Create API-only users that only have access to Customer-defined Fleet API endpoints (#42352)
fleetdm/fleet: docs/REST API/rest-api.md

- [Delete invite](#delete-invite)
- [Verify invite](#verify-invite)
- [Update invite](#update-invite)

> - [List API endpoints for API-only user permissions](#list-api-endpoints-for-api-only-user-permissions)

Author: ```suggestion
- [Create API-only user](#create-api-only-user)
- [List API endpoints for API-only user permissions](#list-api-endpoints-for-api-only-user-permissions)
```

}
```

### Create API-only user

Creates an API-only user that does not have access to the UI.

`POST /api/v1/fleet/users/api_only`

| Name                    | Type    | In    | Description |
| :---------------------- | :------ | :---- | :---------- |

> | password    | string  | body | The user's password (required for non-SSO users).                                        

Author: ```suggestion
| name    | string  | body | The display name for the API-only user. |                                         
```

| global_role | string | body | The role assigned to the user. If `global_role` is specified, `fleets` cannot be specified. For more information, see [manage access](https://fleetdm.com/docs/using-fleet/manage-access). |
| fleets      | array   | body | _Available in Fleet Premium_. The fleets and respective roles assigned to the user. Should contain an array of objects in which each object includes the fleet's `id` and the user's `role` on each fleet. If `fleets` is specified, `global_role` cannot be specified. For more information, see [manage access](https://fleetdm.com/docs/using-fleet/manage-access). |

> | api_endpoints | array   | body | _Available in Fleet Premium_. A list of `id`s of API endpoints this user will have access to. For available endpoints, see [List API endpoints for API-only user permissions](#list-api-endpoints-for-api-only-user-permissions). |

Reviewer: @rachaelshaw 
When omitted or empty it means access to ALL, right?

And on the UI we probably want to disable the `Save` button if the user selected `Specific API endpoints` but didn't choose any.

Author: Cood catch 👍 We should make the user select at least one endpoint or "All" before saving
Would it match existing patterns to have this be `null` if not setting specific endpoints?

Author: ^ @juan-fdz-hawa

If `api_endpoints` is specified, these do not grant additional permissions otherwise forbidden by the user's `role`.

#### Example

`POST /api/v1/fleet/users/admin`

##### Request body

```json
{
  "name": "Jane Doe",
  "fleets": [
    {
      "id": 2,
      "role": "observer"
    },
    {
      "id": 3,
      "role": "maintainer"
    }
  ], 

>   "api_endpoints": [1,5,7,32]

Reviewer: @rachaelshaw 
I put as a goal on the implementation to make these `id`s "stable".
Meaning that for all users/customers the ids map to the same endpoints (to ease troubleshooting and reproducing issues).

}
```

##### Default response

`Status: 200`

```json
{
  "user": {
    "created_at": "0001-01-01T00:00:00Z",
    "updated_at": "0001-01-01T00:00:00Z",
    "id": 5,
    "name": "Jane Doe",

>     "email": "janedoe+randomlygeneratedstring@example.com",

Author: yeah that's the admin's email

      },
      {
        "id": 3,
        "role": "maintainer"
      }
    ], 
    "api_endpoints": [1,5,7,32]
  },
  "token": "{API key}"
}
```

### List API endpoints for API-only user permissions

 _Available in Fleet Premium._

Lists Fleet REST API endpoints that an API-only user can be granted access to.

`GET /api/v1/fleet/rest_api`

| Name                    | Type    | In    | Description |
| :---------------------- | :------ | :---- | :---------- |
| query                   | string  | query | Search query keywords. Searchable fields include `display_name` and `path`. |

Searching by path ignores the naming of path parameters that are specified with `:` , e.g. `:id`. So searching `/hosts/:id/report` is the same as searching `/hosts/:host_id/report`.

Experimental endpoints are excluded from the results, since they are not for use in automated workflows.

#### Example

`GET /api/v1/fleet/rest_api?query=get%20host%20by%20identifier`
or
`GET /api/v1/fleet/rest_api?query=%2Fapi%2Fv1%2Ffleet%2Fhosts%2Fidentifier%2F%3Ahost_identifier`

##### Default response

`Status: 200`

```json
{
  "api_endpoints": [
    {
      "id": 123,
      "display_name": "Get host by identifier",

>       "protocol": "GET",

Reviewer 2: @rachaelshaw @lucasmrod Can this be named `method` instead? `protocol` is something else (HTTP, SSH, etc).

Author: ```suggestion
      "method": "GET",
```
